src/stub_open.c

```diff
@@ -14,21 +15,132 @@
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     r+      │ O_RDWR                        │ FA_READ | FA_WRITE                    │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     w+      │ O_RDWR | O_CREAT | O_TRUNC    │ FA_CREATE_ALWAYS | FA_WRITE | FA_READ │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     a+      │ O_RDWR | O_CREAT | O_APPEND   | FA_OPEN_APPEND | FA_WRITE | FA_READ   │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+*/
+static int _map_oflags(int oflag) {
     int result = 0;
-    if (oflag & O_RDONLY) result |= FA_READ;
-    if (oflag & O_WRONLY) result |= FA_WRITE;
-    if (oflag & O_APPEND) result |= FA_OPEN_APPEND;
-    if (oflag & O_CREAT & O_TRUNC) {
-        result |= FA_CREATE_ALWAYS;
-    } else if (oflag & O_CREAT) {
-        result |= FA_CREATE_NEW;
+
+    switch(oflag) {
+        // r
+        case O_RDONLY:
+            result = FA_READ;
+            break;
+        // w
+        case (O_WRONLY | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE;
+            break;
+        // a
+        case (O_WRONLY | O_CREAT | O_APPEND ):
+            result = FA_OPEN_APPEND | FA_WRITE;
+            break;
+        // r+
+        case (O_RDWR ):
+            result = FA_READ | FA_WRITE;
+            break;
+        // w+
+        case (O_RDWR | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE | FA_READ;
+            break;
+        // a+
+        case (O_RDWR | O_CREAT | O_APPEND):
+            result = FA_OPEN_APPEND | FA_WRITE | FA_READ;
+            break;
+        default:
+            errno = EINVAL;
+            result = -1;
+            break;
     }
+
     return result;
 }
 
+static void _set_errno(int result) {
+    switch(result) {
+        case FSYS_ERR_DISK_ERR:
+        case FSYS_ERR_INT_ERR:
+        case FSYS_ERR_NOT_READY:
+        case FSYS_ERR_NOT_ENABLED:
+        case FSYS_ERR_NO_FILESYSTEM:
+            errno = ENODEV;
+            break;
+        case FSYS_ERR_NO_FILE:
+        case FSYS_ERR_NO_PATH:
+            errno = ENOENT;
+            break;
+        case FSYS_ERR_INVALID_NAME:
+        case FSYS_ERR_INVALID_OBJECT:
+        case FSYS_ERR_INVALID_DRIVE:
+        case FSYS_ERR_INVALID_PARAMETER:
+            errno = EINVAL;
+            break;
+        case FSYS_ERR_DENIED:
+        case FSYS_ERR_WRITE_PROTECTED:
+            errno = EACCES;
+            break;
+        case FSYS_ERR_EXIST:
+            errno = EEXIST;
+            break;
+        case FSYS_ERR_TOO_MANY_OPEN_FILES:
+            errno = ENFILE;
+            break;
+        case FSYS_ERR_LOCKED:
+            errno = ETXTBSY;
+            break;
+        case FSYS_ERR_MKFS_ABORTED:
+        case FSYS_ERR_TIMEOUT:
+        case FSYS_ERR_NOT_ENOUGH_CORE:
+            errno = EFAULT;
+            break;
+    }
```

Comment:
I think there should be a default case here, not sure what number it should assign though. This is in case more errors are added (or we do not have complete cover) so we get something "sensible"  into `errno`.